daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -613,4 +604,164 @@ object Transaction {
+                  key1.forall(_ => key1 == key2) =>
+              loop(rest1, rest2, stack)
+            case (
+                exe1 @ Node.NodeExercises(
+                  targetCoid1,
+                  templateId1,
+                  choiceId1,
+                  optLocation1 @ _,
+                  consuming1,
+                  actingParties1,
+                  chosenValue1,
+                  stakeholders1,
+                  signatories1,
+                  choiceObservers1,
+                  controllersDifferFromActors1,
+                  children1 @ _,
+                  exerciseResult1,
+                  key1,
+                  byKey1 @ _,
+                ),
+                exe2 @ Node.NodeExercises(
+                  targetCoid2,
+                  templateId2,
+                  choiceId2,
+                  optLocation2 @ _,
+                  consuming2,
+                  actingParties2,
+                  chosenValue2,
+                  stakeholders2,
+                  signatories2,
+                  choiceObservers2,
+                  controllersDifferFromActors2,
+                  children2 @ _,
+                  exerciseResult2,
+                  key2,
+                  byKey2 @ _,
+                )
+                ) =>
+              // Exercise result are check after having checking the children
+              if (targetCoid1 === targetCoid2 && templateId1 == templateId2 && choiceId1 == choiceId2 &&
+                consuming1 == consuming2 && actingParties1 == actingParties2 && chosenValue1 === chosenValue2 &&
+                stakeholders1 == stakeholders2 && signatories1 == signatories2 && choiceObservers1 == choiceObservers2 &&
+                controllersDifferFromActors1 == controllersDifferFromActors2 &&
+                exerciseResult1.fold(true)(_ => exerciseResult1 === exerciseResult2) &&
+                key1.fold(true)(_ => key1 === key2)) {
+                loop(
+                  children1.iterator.toStream,
+                  children2.iterator.toStream,
+                  (nid1, exe1, rest1, nid2, exe2, rest2) :: stack
+                )
+              } else
+                fail(Some(nid1), Some(nid2))
+            case (
+                Node.NodeLookupByKey(templateId1, optLocation1 @ _, key1, result1),
+                Node.NodeLookupByKey(templateId2, optLocation2 @ _, key2, result2)
+                ) if templateId1 == templateId2 && key1 === key2 && result1 === result2 =>
+              loop(rest1, rest2, stack)
+            case _ => fail(Some(nid1), Some(nid2))
+          }
+
+        case (Stream.Empty, Stream.Empty) =>
+          stack match {
+            case (nid1, exe1, nids1, nid2, exe2, nids2) :: rest =>
+              if (exe1.exerciseResult === exe2.exerciseResult)
+                loop(nids1, nids2, rest)
+              else
+                fail(Some(nid1), Some(nid2))
+            case Nil =>
+              Right(())
+          }
+
+      }
+
```

Comment:
You probably also want to have a case here if the streams have different lengths. Otherwise we get an exception instead of a `ReplayMismatch`.

Reply:
good catch. I would have expected incomplete pattern matching warning.